pkg/collector/collector.go

```diff
@@ -0,0 +1,534 @@
+				keyUpgradeConfigName,
+			}, nil),
+		upgradeHealthCheckFailed: prometheus.NewDesc(
+			prometheus.BuildFQName(MetricsNamespace, subSystemCluster, "health_check_failed"),
+			helpUpgradeHealthCheckFailed,
+			[]string{
+				keyUpgradeConfigName,
+				keyState,
+			}, nil),
+		scalingFailed: prometheus.NewDesc(
+			prometheus.BuildFQName(MetricsNamespace, subSystemUpgrade, "scaling_failed"),
+			helpScalingFailed,
+			[]string{
+				keyDimension,
+			}, nil),
+		clusterVerificationFailed: prometheus.NewDesc(
+			prometheus.BuildFQName(MetricsNamespace, subSystemCluster, "verification_failed"),
+			helpClusterVerificationTimeout,
+			[]string{
+				keyUpgradeConfigName,
+				keyPhase,
+			}, nil),
+		controlPlaneTimeout: prometheus.NewDesc(
+			prometheus.BuildFQName(MetricsNamespace, subSystemUpgrade, "control_plane_timeout"),
+			helpControlPlaneTimeout,
+			[]string{
+				keyUpgradeConfigName,
+				keyPhase,
+			}, nil),
+		workerTimeout: prometheus.NewDesc(
+			prometheus.BuildFQName(MetricsNamespace, subSystemUpgrade, "worker_timeout"),
+			helpWorkerTimeout,
+			[]string{
+				keyUpgradeConfigName,
+				keyVersion,
+			}, nil),
+		nodeDrainFailed: prometheus.NewDesc(
+			prometheus.BuildFQName(MetricsNamespace, subSystemUpgrade, "node_drain_failed"),
+			helpNodeDrainFailed,
+			[]string{
+				keyNodeName,
+			}, nil),
+		upgradeWindowBreached: prometheus.NewDesc(
+			prometheus.BuildFQName(MetricsNamespace, subSystemUpgrade, "window_breached"),
+			helpUpgradeWindowBreached,
+			[]string{
+				keyUpgradeConfigName,
+			}, nil),
+	}
+
+	//	osdMetrics := &OSDUpgradeMetrics{
+	//		providerSyncFailed: prometheus.NewDesc(
+	//			prometheus.BuildFQName(MetricsNamespace, subSystemUpgradeConfig, "sync_failed"),
+	//			"Boolean indicating if the sync with UpgradeConfig provider failed",
+	//			[]string{
+	//				keyUpgradeConfigName,
+	//			}, nil),
+	//	}
+	osdMetrics := &OSDUpgradeMetrics{
+		notificationEventSentFailed: prometheus.NewDesc(
+			prometheus.BuildFQName(MetricsNamespace, subSystemNotification, "event_sent_failed"),
+			helpNotificationEventSentFailed,
+			[]string{
+				keyUpgradeConfigName,
+				keyState,
+				keyVersion,
+			}, nil),
+	}
+	aroMetrics := &AROUpgradeMetrics{}
+
+	return &UpgradeCollector{
+		upgradeConfigManager,
+		cv.NewCVClient(c),
+		coreMetrics,
+		osdMetrics,
+		aroMetrics,
+	}, nil
+}
+
+// Describe implements the prometheus.Collector interface.
+func (uc *UpgradeCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- uc.coreUpgradeMetrics.upgradeState
+	ch <- uc.coreUpgradeMetrics.configInvalid
+	ch <- uc.coreUpgradeMetrics.upgradeHealthCheckFailed
+	ch <- uc.coreUpgradeMetrics.scalingFailed
+	ch <- uc.coreUpgradeMetrics.clusterVerificationFailed
+	ch <- uc.coreUpgradeMetrics.controlPlaneTimeout
+	ch <- uc.coreUpgradeMetrics.workerTimeout
+	ch <- uc.coreUpgradeMetrics.nodeDrainFailed
+	ch <- uc.coreUpgradeMetrics.upgradeWindowBreached
+
+	ch <- uc.osdUpgradeMetrics.notificationEventSentFailed
```

Comment:
Not really sure if you want to tackle it in this PR or wait until M7 but I feel there is lots of opportunity here to start thinking in advance about how this works when we have different upgraders around.

There are some early provisions in this PR for ARO collectors, but none of the code yet is looking at the upgrade type when collecting and sending various metrics.

ie. how will this function and (more specifically) the associated `collectUpgradeMetrics` behave when there are ARO and OSD upgraders?

ARO's metrics gathering doesn't need to be solved yet, but in terms of this PR I guess what I'm thinking is.. can some of these functions be changed in advance of ARO so that we only collect and send OSD related metrics if the upgradeconfig's type actually is "OSD"? (forgive me if that is actually happening and I've missed it). Solving that is half the battle of this issue and can be done ahead of time.